### PR TITLE
StratifiedRules: Set filter

### LINKED ISSUE #2642: StratifiedRules: Set filter
Add new rule for `{x \in S: p}`.

### PR DESCRIPTION
**Pull request by Kukovec**
- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] [Entries added to `./unreleased/`][changelog format] for any new functionality

closes #2642 

This PR updates SetFilterRule to the new approach (by introducing a new rule, due to incompatible interfaces).

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

**Comment by shonfeder** [comment]
Could you expand the PR or issue to help give context on this changeset? As things stand, I can't tell what is motivating this PR. The linked issue just says

> Add new rule for `{x \in S: p}`

But there is no description of what this is being rewritten into. Given that this is using the new, stratified design, I was expecting that the PR would be migrating an existing rule. But that isn't the case, so I'm also not clear why we are introducing new rules at this point rather than paying down the tech debt of the existing rewriting system.

Thank you in advance for the clarification!

**Comment by Kukovec** [comment]
> Could you expand the PR or issue to help give context on this changeset? As things stand, I can't tell what is motivating this PR. The linked issue just says
> 
> > Add new rule for `{x \in S: p}`
> 
> But there is no description of what this is being rewritten into. Given that this is using the new, stratified design, I was expecting that the PR would be migrating an existing rule. But that isn't the case, so I'm also not clear why we are introducing new rules at this point rather than paying down the tech debt of the existing rewriting system.
> 
> Thank you in advance for the clarification!

Since the existing rules and the new stratified rules have an incompatible interface (they don't use `SymbState`), you can't just rewrite them in place, so "migrating" a rule is really just a new class based on the old file, but which lives separately. The convention is, that for every `FooRule`, there is/will be a class `FooStratifiedRule` using `Rewriter(Impl)` instead of `SymbStateRewriter(Impl)`.

TLDR, you can see this PR as porting `SetFilterRule` to the new approach.

**Comment by thpani** [comment]
> TLDR, you can see this PR as updating `SetFilterRule` to the new approach.

This should go in the PR description.
Fishing for context only delays the review process.

**Review by konnov — APPROVED** [review]
Looks like a good start!

**Review by thpani — APPROVED** [review]
Code looks good, please update the PR description.

File: tla-bmcmt/src/test/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/SetRewriterTest.scala [hunk 1]
```diff
@@ -123,4 +121,98 @@ class SetRewriterTest extends AnyFunSuite with BeforeAndAfterEach {
...
+    val elems = Seq(101, 102, 103).map(new ArenaCell(_, CellT.fromType1(IntT1)))
+
+    val arena = PureArena.initial.appendCell(setCell).appendCellSeq(elems: _*)
+
+    val arenaNoDups = arena.appendHas(setCell, elems.map(FixedElemPtr): _*)
+    val arenaWithDups = arena.appendHas(setCell, (elems ++ elems).map(FixedElemPtr): _*)
+
+    // TODO: since other rules aren't implemented yet, we can't do interesting filters
+    // TODO: (e.g. filtering an int range by a midpoint). The best we can do now is const-filter with T/F
+    // TODO: revisit these tests once more rules are added
```
[the remark this answers is not in the document]
Comment (Kukovec) [reply tla-bmcmt/src/test/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/SetRewriterTest.scala:164]: https://github.com/informalsystems/apalache/issues/2652
